runner/logs.go

```diff
@@ -1,57 +1,108 @@
-		if globalCollectionOpts.TestRun {
-			logger.PrintError("  Failed - Log Insights feature not available on this pganalyze plan, or log data limit exceeded. You may need to upgrade, see https://pganalyze.com/pricing")
-		} else {
-			logger.PrintVerbose("Skipping log data: Feature not available on this pganalyze plan, or log data limit exceeded")
+	for _, server := range servers {
+		if server.Config.DisableLogs {
+			continue
+		}
+		if server.Config.LogLocation != "" || server.Config.LogDockerTail != "" || server.Config.LogSyslogServer != "" {
+			hasAnyLogTails = true
+		} else if server.Config.AwsDbInstanceID != "" {
+			hasAnyLogDownloads = true
 		}
-		return newLogState, false, nil
 	}
 
-	transientLogState, persistedLogState, err := input.DownloadLogs(server, server.LogPrevState, globalCollectionOpts, logger)
-	if err != nil {
-		transientLogState.Cleanup()
-		return newLogState, false, errors.Wrap(err, "could not collect logs")
+	var parsedLogStream chan state.ParsedLogStreamItem
+	if hasAnyLogTails || hasAnyHeroku || hasAnyGoogleCloudSQL || hasAnyAzureDatabase {
+		parsedLogStream = setupLogStreamer(ctx, wg, globalCollectionOpts, logger, servers, nil, stream.LogTestNone)
+	}
+	if hasAnyLogTails {
+		selfhosted.SetupLogTails(ctx, wg, globalCollectionOpts, logger, servers, parsedLogStream)
+	}
+	if hasAnyHeroku {
+		heroku.SetupHttpHandlerLogs(ctx, wg, globalCollectionOpts, logger, servers, parsedLogStream)
+	}
+	if hasAnyGoogleCloudSQL {
+		google_cloudsql.SetupLogSubscriber(ctx, wg, globalCollectionOpts, logger, servers, parsedLogStream)
+	}
+	if hasAnyAzureDatabase {
+		azure.SetupLogSubscriber(ctx, wg, globalCollectionOpts, logger, servers, parsedLogStream)
 	}
 
-	newLogState = persistedLogState
-
-	err = output.UploadAndSendLogs(server, grant, globalCollectionOpts, logger, transientLogState)
-	if err != nil {
-		transientLogState.Cleanup()
-		return newLogState, false, errors.Wrap(err, "failed to upload/send logs")
+	if hasAnyLogDownloads {
+		setupLogDownloadForAllServers(ctx, wg, globalCollectionOpts, logger, servers)
 	}
+}
 
-	transientLogState.Cleanup()
-	return newLogState, true, nil
+func setupLogDownloadForAllServers(ctx context.Context, wg *sync.WaitGroup, globalCollectionOpts state.CollectionOpts, logger *util.Logger, servers []*state.Server) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		ticker := time.NewTicker(LogDownloadInterval)
+
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				var innerWg sync.WaitGroup
+
+				if !globalCollectionOpts.CollectLogs {
+					return
+				}
+
+				for _, server := range servers {
+					if server.Config.DisableLogs || (server.Grant.Valid && !server.Grant.Config.EnableLogs) {
+						continue
+					}
+
+					if server.Config.AwsDbInstanceID == "" {
+						continue
+					}
+
+					innerWg.Add(1)
+					go downloadLogsForServerWithLocksAndCallbacks(&innerWg, server, globalCollectionOpts, logger)
+				}
+
+				innerWg.Wait()
```

Comment:
This lets us avoid downloading new log files if previous log files are still being processed, right? Nice.

Reply:
Yeah, right now we always finish the work for all servers before scheduling the work again (the same is done for other snapshot data). This was already in place before for log downloads as well, FWIW.

Not sure if its the right decision, as it could lead to a slow server causing all other servers on a collector to not have collections at regular intervals. But for now this is consistent this way, and we may re-evaluate this at some point.